lychee-bin/src/archive/mod.rs

```diff
@@ -0,0 +1,104 @@
+use anyhow::anyhow;
+use std::str::FromStr;
+
+use http::StatusCode;
+use reqwest::{Error, Url};
+use serde::{Deserialize, Deserializer, Serialize};
+
+#[derive(Debug, Serialize, Eq, Hash, PartialEq)]
+pub(crate) struct Suggestion {
+    pub(crate) original: Url,
+    pub(crate) suggestion: Url,
+}
+
+#[derive(Debug, Deserialize, Default, Clone)]
+pub(crate) enum Archive {
+    #[default]
+    WaybackMachine,
+}
+
+impl FromStr for Archive {
+    type Err = anyhow::Error;
+    fn from_str(archive: &str) -> Result<Self, Self::Err> {
+        match archive.to_lowercase().as_str() {
+            "wayback" => Ok(Archive::WaybackMachine),
+            _ => Err(anyhow!("Unknown archive {}", archive)),
+        }
+    }
+}
+
+impl Archive {
+    pub(crate) async fn get_link(&self, original: &Url) -> Result<Option<Url>, Error> {
+        let function = match self {
+            Archive::WaybackMachine => get_wayback_link,
+        };
+
+        function(original).await
+    }
+}
+
+async fn get_wayback_link(url: &Url) -> Result<Option<Url>, Error> {
+    let mut archive_url = Url::parse("https://archive.org/wayback/available").unwrap();
+    archive_url.set_query(Some(&format!("url={url}")));
+
+    let response = reqwest::get(archive_url)
+        .await?
+        .json::<InternetArchiveResponse>()
+        .await?;
+
+    Ok(response
+        .archived_snapshots
+        .closest
+        .map(|closest| closest.url))
+}
+
+#[derive(Debug, Deserialize, Eq, PartialEq)]
+pub(crate) struct InternetArchiveResponse {
```

Comment:
is this and next two structs used for wayback machine only?

they can be organized in a sub module.